modules/assim.sequential/R/sda.enkf_MultiSite.R

```diff
@@ -142,338 +151,464 @@ sda.enkf.multisite <- function(settings, obs.mean, obs.cov, Q = NULL, restart=F,
-        PEcAn.logger::logger.warn ("\n --------------Analysis Cov ----------- \n")
-        print(enkf.params[[t]]$Pa)
-        PEcAn.logger::logger.warn ("\n ------------------------------------------------------\n")
+      ###  preparing OBS                                                    ###
+      ###-------------------------------------------------------------------###---- 
+      if (any(obs)) {
+        if (control$debug) browser()
+        #Making R and Y
+        Obs.cons <- Construct.R(site.ids, var.names, obs.mean[[t]], obs.cov[[t]])
+        
+        Y <- Obs.cons$Y
+        R <- Obs.cons$R
+        
+        if (length(Y) > 1) {
+          PEcAn.logger::logger.info("The zero variances in R and Pf is being replaced by half and one fifth of the minimum variance in those matrices respectively.")
+          diag(R)[which(diag(R)==0)] <- min(diag(R)[which(diag(R) != 0)])/2
+        }
+        # making the mapping oprator
+        H <- Construct.H.multisite(site.ids, var.names, obs.mean[[t]])
+        
+        ###-------------------------------------------------------------------###
+        ### Analysis                                                          ###
+        ###-------------------------------------------------------------------###----
+        
+        tic(paste0("Analysis for cycle = ", t))
+        
+        if(processvar == FALSE){an.method<-EnKF.MultiSite  }else{    an.method<-GEF.MultiSite   }  
+        
+        #-analysis function
+        enkf.params[[obs.t]] <- Analysis.sda(
+          settings,
+          FUN = an.method,
+          Forecast = list(Q = Q, X = X),
+          Observed = list(R = R, Y = Y),
+          H = H,
+          extraArg = list(
+            aqq = aqq,
+            bqq = bqq,
+            t = t,
+            nitr.GEF = nitr.GEF,
+            nthin = nthin,
+            nburnin = nburnin,
+            censored.data = censored.data
+          ),
+          choose = choose,
+          nt = nt,
+          obs.mean = obs.mean,
+          obs.cov = obs.cov,
+          site.ids = site.ids,
+          blocked.dis = blocked.dis,
+          distances = distances
+        )
+        tic(paste0("Preparing for Adjustment for cycle = ", t))
+        #Forecast
+        mu.f <- enkf.params[[obs.t]]$mu.f
+        Pf <- enkf.params[[obs.t]]$Pf
+        #Analysis
+        Pa <- enkf.params[[obs.t]]$Pa
+        mu.a <- enkf.params[[obs.t]]$mu.a
+        #extracting extra outputs
+        if (control$debug) browser()
+        if (processvar) {
+          aqq <- enkf.params[[obs.t]]$aqq
+          bqq <- enkf.params[[obs.t]]$bqq
+        }
+        # Adding obs elements to the enkf.params
+        #This can later on help with diagnostics
+        enkf.params[[obs.t]] <- c(enkf.params[[obs.t]], R = list(R))
+        enkf.params[[obs.t]] <- c(enkf.params[[obs.t]], Y = list(Y))
+        enkf.params[[obs.t]] <- c(enkf.params[[obs.t]], RestartList = list(restart.list %>% setNames(site.ids) ))
```

Comment:
Seems like it would be much more efficient to add all 3 of these things at once, rather than 3 separate lines. If enkf.params is large, then this is a lot extra copying going on within implicit temporary objects.